### PR TITLE
Add OmnitelK Exporter Client implementation

### PR DESCRIPTION
**Pull request by tigrannajaryan**
Added multi-stream client implementation. This commit
includes partial test coverage, but need more tests.

TODO: add tests for server closing connection, server closing stream,
server returning non SUCCESS responses.

Task: A0-1611

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@pjanotti @owais please have a look.

File: exporter/omnitelk/test_helpers.go [hunk 1]
```diff
@@ -118,3 +120,18 @@ func GenRandByteString(len int) string {
 	}
 	return string(b)
 }
+
+// GetAvailableLocalAddress finds an available local port and returns an endpoint
+// describing it. The port is available for opening when this function returns
+// provided that there is no race by some other code to grab the same port
+// immediately.
+func GetAvailableLocalAddress() string {
+	ln, err := net.Listen("tcp", ":0")
```
Comment (pjanotti) [inline exporter/omnitelk/test_helpers.go]: Use `localhost:0`to avoid the pop-up on locked down macOS.

File: exporter/omnitelk/mock_server.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+func runServer(srv *mockServer, listenAddress string) {
+
+	log.Printf("Server: listening on %s", listenAddress)
+
+	if srv.config == nil {
+		srv.config = &omnitelpb.ShardingConfig{}
+	}
+
+	//totalSpans := 0
```
Comment (pjanotti) [inline exporter/omnitelk/mock_server.go]: nit: remove commented lines, or if desired for debugging make it behavior a config.

File: exporter/omnitelk/mock_server.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+func (s *gRPCServer) Export(stream omnitelpb.OmnitelK_ExportServer) error {
+	for {
+		// Wait for request from client.
+		request, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return nil
```
Comment (pjanotti) [inline exporter/omnitelk/mock_server.go]: `return err` no?

File: exporter/omnitelk/mock_server.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+		// Process received request.
+		s.onReceive(request)
+
+		// Send response to client.
+		response := &omnitelpb.ExportResponse{
+			Id:             request.Id,
+			ResultCode:     s.srv.nextResponseCode,
+			ShardingConfig: s.srv.nextResponseShardingConfig,
+		}
+		stream.Send(response)
```
Comment (pjanotti) [inline exporter/omnitelk/mock_server.go]: check possible error

File: exporter/omnitelk/client_impl_test.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+func TestConnectCancellation(t *testing.T) {
+	c := NewClient(zap.NewNop())
+
+	cancelCh := make(chan interface{})
+	done := make(chan interface{})
+
+	var err error
+	go func() {
+		err = c.Connect(ConnectionOptions{
+			Endpoint: "0.0.0.0:0",
```
Comment (pjanotti) [inline exporter/omnitelk/client_impl_test.go]: nit: localhost to avoid pop up on macOS

File: exporter/omnitelk/client_impl_test.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	cs.failedRecords = append(cs.failedRecords, failedRecords)
+	cs.failureCodes = append(cs.failureCodes, code)
+}
+
+func (cs *clientSink) getResponseToRecords() []*omnitelpb.EncodedRecord {
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	return cs.responseToRecords
```
Comment (pjanotti) [inline exporter/omnitelk/client_impl_test.go]: nit: get a copy

File: exporter/omnitelk/client_impl_test.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+func (cs *clientSink) getResponses() []*omnitelpb.ExportResponse {
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	return cs.responses
```
Comment (pjanotti) [inline exporter/omnitelk/client_impl_test.go]: nit: get a copy

File: exporter/omnitelk/client_impl_test.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		// Check that all responses were success.
+		responses := clientSink.getResponses()
+		for i := 0; i < len(responses); i++ {
+			response := responses[i]
+			assert.True(t, response.Id != 0)
+			assert.EqualValues(t, omnitelpb.ExportResponse_SUCCESS, response.ResultCode)
+			assert.Nil(t, response.ShardingConfig)
+		}
+
+		server.Stop()
```
Comment (pjanotti) [inline exporter/omnitelk/client_impl_test.go]: nit: although this test has only one `require` that can skip the stop it seems better to have `defer server.Stop()` right after the call to `setupClientServer`

File: exporter/omnitelk/client_impl_test.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+// the PartitionKey field.
+type byPartitionKey []*omnitelpb.EncodedRecord
+
+func (a byPartitionKey) Len() int      { return len(a) }
+func (a byPartitionKey) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a byPartitionKey) Less(i, j int) bool {
+	return a[i].PartitionKey < a[j].PartitionKey
+}
+
+func TestSendRequest(t *testing.T) {
```
Comment (pjanotti) [inline exporter/omnitelk/client_impl_test.go]: nit: matter of personal preference but I tend to prefer the test functions on the top before the definition of the helpers.

File: exporter/omnitelk/client_impl_test.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+	// Connect to the server
+	err := client.Connect(ConnectionOptions{
+		Endpoint: endpoint,
+	}, cancelCh)
+
+	// Make sure connection succeeded.
+	assert.Nil(t, err)
+
+	server.Stop()
```
Comment (pjanotti) [inline exporter/omnitelk/client_impl_test.go]: nit: `defer server.Stop()` right after the runServer goroutine.

File: exporter/omnitelk/client.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+	omnitelpb "github.com/Omnition/omnition-opentelemetry-service/exporter/omnitelk/gen"
+)
+
+// client allows to connect to a server, get sharding config and send encoded data.
+type client interface {
+	// Connect to the server endpoint using specified number of concurrent streams.
+	// Connect must block until it succeeds or fails (and return error in that case).
+	// If caller needs to interrupt a blocked Connect call the caller must close
+	// "cancelCh" channel, in that case Connect should return as soon as possible.
```
Comment (pjanotti) [inline exporter/omnitelk/client.go]: Side note: I don't mind using the `cancelCh` but the typical thing in this case is to use a context for cancellation.

File: exporter/omnitelk/client.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	Connect(options ConnectionOptions, cancelCh chan interface{}) error
+
+	// GetShardingConfig returns a sharding config from the server. May be called
+	// only after Connect succeeds.
+	GetShardingConfig() (*omnitelpb.ShardingConfig, error)
+
+	// Send an encoded record to the server. The record must be encoded for the shard
+	// that is passed as the second parameter (record's partition key must be in the
+	// has key range of the shard).
```
Comment (pjanotti) [inline exporter/omnitelk/client.go]:
```suggestion
	// key range of the shard).
```

File: exporter/omnitelk/client.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	// This function will block if it wants to apply backpressure otherwise it may
+	// return as soon as the record is queued for delivery.
+	// The result of sending will be reported via OnSendResponse or OnSendFail
+	// callbacks.
+	Send(record *omnitelpb.EncodedRecord, shard *omnitelpb.ShardDefinition)
+}
+
+// ConnectionOptions to use for the client.
+type ConnectionOptions struct {
+	// mockServer's address and port.
```
Comment (pjanotti) [inline exporter/omnitelk/client.go]: nit: fix comment

File: exporter/omnitelk/client.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	// Also reopen the stream after specified count of requests are sent.
+	StreamReopenRequestCount uint32
+
+	// Callback called when a response is received regarding previously sent records.
+	// Called asynchronously sometime after Send() successfully sends the record.
+	OnSendResponse func(responseToRecords []*omnitelpb.EncodedRecord, response *omnitelpb.ExportResponse)
+
+	// Callback called if the records cannot be sent for whatever reason (e.g. the
+	// records cannot be serialized).
+	// TODO: move FailureCode definition to a common file.
```
Comment (pjanotti) [inline exporter/omnitelk/client.go]: nit: isn't this TODO already done below? Or do you mean out of OmnitelK?

File: exporter/omnitelk/client_impl_test.go (hunk 6 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/client_impl_test.go]: I think we don't need to copy. Return value points to a slice of immutable EncodedRecords. responseToRecords may be modified later by appending but that has no effect on the returned slice. It should be thus safe to return the slice reference as is. Do I misunderstand something?

File: exporter/omnitelk/client.go (hunk 11 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/client.go]: I will need to change a lot for this and I already have the next PR depending on this which will require even more changes. Let me re-think it later, I'd like to merge it as is. Using channels for cancellation is also quite typical Go idiom: https://blog.golang.org/pipelines

**Comment by tigrannajaryan** [comment]
@pjanotti thanks for the review. PTAL.

File: exporter/omnitelk/mock_server.go (hunk 2 above)
Comment (pjanotti) [reply exporter/omnitelk/mock_server.go]: @tigrannajaryan missed this one?

File: exporter/omnitelk/client_impl_test.go (hunk 6 above)
Comment (pjanotti) [reply exporter/omnitelk/client_impl_test.go]: If the slice has capacity and an append happens after the function returned it won't reflect the state when the function was called. I didn't see it being used in a context with such issue but since this a test function and performance is not as much of concern I think it is better to add it.

Comment (tigrannajaryan) [reply exporter/omnitelk/client_impl_test.go]: I don't think what you are saying is true.
Slices internally are 3 fields internally: pointer to underlying array, len and cap. When you copy a slice variable all 3 fields are copied and both slices point to the same underlying array. When you append an item to a slice it either will append in place if there is enough capacity or will reallocate underlying array. In either case the variable containing the copy is not affected and continues to point to the old content. Source: https://blog.golang.org/go-slices-usage-and-internals

Try this:
```
    s1 := []int{1, 2, 3}
    s2 := s1
    s1 = append(s1, 4)
    fmt.Printf("%08d\n", s1)
    fmt.Printf("%08d\n", s2)
```

So if `cs.responseToRecords` is modified after the function returns it will have no effect on the returned value, which will correctly reflect the state of the slice at the time of the call.

File: exporter/omnitelk/mock_server.go (hunk 2 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/mock_server.go]: Do you mind if I merge as is? I have this already fixed in a subsequent PR and want to avoid rebase conflicts.

File: exporter/omnitelk/client_impl.go [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+
+			pr := elem.Value.(pendingRequest)
+
+			// Check request deadline.
+			if pr.deadline.Before(now) {
+				// We are past deadline, need to fail the request.
+				c.pendingAckList.Remove(elem)
+				delete(c.pendingAckMap, pr.request.Id)
+				c.pendingAckMutex.Unlock()
+				c.client.options.OnSendFail(pr.request.Records, FailedNotRetryable)
```
Comment (pjanotti) [inline exporter/omnitelk/client_impl.go]: timeout should be retriable, if not probably deserves a comment of why not.

File: exporter/omnitelk/client_impl.go [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+// callbacks.
+func (c *Client) Send(record *omnitelpb.EncodedRecord, shard *omnitelpb.ShardDefinition) {
+	request := &omnitelpb.ExportRequest{
+		Records: []*omnitelpb.EncodedRecord{record},
+		Shard:   shard,
+	}
+
+	if c.options.StreamCount == 1 {
+		// One stream, it is faster to send request directly, bypassing the queue.
+		c.clientStreams[0].sendRequest(request)
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/client_impl.go]: Good catch! I already fixed this the next PR: https://github.com/Omnition/omnition-opentelemetry-service/pull/13/filesin

File: exporter/omnitelk/client_impl.go [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+			// an error to RP call. We will need to reconnect.
+			// This can probably also happen when we close the stream, so need
+			// to be careful not to reconnect in that case.
+			return
+		}
+
+		c.pendingAckMutex.Lock()
+		elem, ok := c.pendingAckMap[response.Id]
+		if !ok {
+			// Response id is not found in pending requests. Should not normally happen,
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/client_impl.go]: There is nothing todo. Comment is improved in https://github.com/Omnition/omnition-opentelemetry-service/pull/13/files

File: exporter/omnitelk/client_impl.go (hunk 15 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/client_impl.go]: I don't think we need to retry on timeout. This indicates something bad hapenning on the serve side, like server being unable to decode the request and getting stuck or forgetting to reply. This behavior can be due to specific data that we are sending, so if we retry we may end up in infinite retries.
This should never happen under normal circumstances, so I believe it is correct to drop in this situation.

File: exporter/omnitelk/client_impl.go [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+	c.pendingAckMutex.Lock()
+	for _, request := range c.pendingAckMap {
+		requests = append(requests, request.Value.(*omnitelpb.ExportRequest))
+	}
+	c.pendingAckMutex.Unlock()
+
+	// Mutex is unlocked asap, now we can send the requests.
+	for _, request := range requests {
+		if err := c.stream.Send(request); err != nil {
+			c.client.options.OnSendFail(request.Records, FailedNotRetryable)
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply exporter/omnitelk/client_impl.go]: This entire retrying logic is pretty poor and needs to be rewritten using a resending queue. I want to do it in a separate PR which will also add proper tests for these cases. Since the code is not activated in any way I think it is OK to merge incomplete implementation and keep iterating on it until we consider it ready for being activated. The reason I am doing it this way is because otherwise I would need to create a giant 5000 line PR to have all functionality with edge cases and tests enabled. Are you OK with this approach.

File: exporter/omnitelk/client_impl_test.go (hunk 6 above)
Comment (pjanotti) [reply exporter/omnitelk/client_impl_test.go]: Sorry, you are correct.

File: exporter/omnitelk/client_impl.go (hunk 15 above)
Comment (pjanotti) [reply exporter/omnitelk/client_impl.go]: Add a comment to describe why not, usually timeout in network operations is treated with bounded number of attempts as you described (you can use the same pending data structure to track number of attempts). That said I think it is fine to go ahead as it is.

Comment (tigrannajaryan) [reply exporter/omnitelk/client_impl.go]: OK, let me address this in a separate PR where I want to do thorough testing of edge cases including timeouts. I don't want to write code for which I don't have tests and don't want to bloat this PR even bigger. Let's move incrementally.

**Comment by tigrannajaryan** [comment]
OK, merging. Will takes notes to fix outstanding issues in future PR.
